admin/RemoveEmptyAccounts.pl

```diff
@@ -0,0 +1,191 @@
+                    FROM place_tag_raw
+                   WHERE place_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM recording_tag_raw
+                   WHERE recording_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM release_tag_raw
+                   WHERE release_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM release_group_tag_raw
+                   WHERE release_group_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM series_tag_raw
+                   WHERE series_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM work_tag_raw
+                   WHERE work_tag_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM artist_rating_raw
+                   WHERE artist_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM event_rating_raw
+                   WHERE event_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM label_rating_raw
+                   WHERE label_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM place_rating_raw
+                   WHERE place_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM recording_rating_raw
+                   WHERE recording_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM release_group_rating_raw
+                   WHERE release_group_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM work_rating_raw
+                   WHERE work_rating_raw.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                    FROM annotation
+                   WHERE annotation.editor = e.id)
+  AND NOT EXISTS (SELECT 1
+                   FROM editor_collection ec
+                  WHERE ec.editor = e.id
+                    AND ((
+                              (type != 5 OR name != 'Attending')
+                              AND
+                              (type != 6 OR name != 'Maybe attending')
+                         )
+                         OR EXISTS (SELECT 1
+                                      FROM editor_collection_event ece
+                                     WHERE ece.collection = ec.id)))
+             AND (   last_login_date IS NULL
+                  OR last_login_date < NOW() - INTERVAL '12 months')
+  SQL
+
+$c_cursor->sql->begin;
+$c_cursor->sql->do("DECLARE cursor NO SCROLL CURSOR FOR $query");
+
+my $pull = sub {
+    $c_cursor->sql->select("FETCH FORWARD $limit FROM cursor");
+};
+while ($pull->()) {
+    $c->sql->begin;
+    my @row;
+    while (@row = $c_cursor->sql->next_row) {
+      if ($dry_run) {
+          print 'Removing account ' . $row[1] . " (dry run)\n";
+      } else {
+          print 'Removing account ' . $row[1] . "\n";
+          # We allow reuse, since these accounts are all unused
+          $c->model('Editor')->delete($row[0], 1);
```

Comment:
Ignoring the actual editor row, is there anything else that'd need deleting besides the defaults collections?  Maybe editor preferences/languages too?  (I'd guess it's not a lot in any case.)

After looking at this again, I think I'd prefer avoiding `$c->model('Editor')->delete` for the prod run and being explicit about which rows we're deleting.  Although the query above looks good, the `delete` method is still dangerous in that it'll delete the editor even if it's not empty.  If there's something we missed, or if the query malfunctions somehow, I'd rather have it fail the deletion (due to existing FKs) than purge everything without question.

Reply:
So, just do all in SQL? I guess that could work, but we should make it so that it does not crash the script if the delete fails. Is there a good way to tell it "on failure, store the username to a file and continue"?

Reply:
IIUC the script shouldn't fail unless there's a bug in the query, or an FK we forgot to delete (but in either case we should fix the script).  The other case would be a transient error (like a timeout), but in that case we can just re-run it, right?

Reply:
Oh, ok, so we'd just let it fail, then fix the script. I was thinking let it delete all it can with no FK clashes, then let's fix, but I guess that'd work too.

Reply:
I think I updated all that was needed, but do make sure I didn't do something stupid, please. I also added a check for no collaborators on collections to the search.